h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -678,114 +679,130 @@ public DataType getKeyType() {
         return map.getKeyType();
     }
 
+    private static final class UncommittedIterator<K, X> extends TMIterator<K, X> {
 
-    private static final class KeyIterator<K> extends TMIterator<K,K> {
-
-        public KeyIterator(TransactionMap<K, ?> transactionMap, K from, K to, boolean includeUncommitted) {
-            super(transactionMap, from, to, includeUncommitted);
-        }
-
-        @Override
-        protected K registerCurrent(K key, VersionedValue data) {
-            return key;
-        }
-    }
+        private final int transactionId;
 
-    private static final class EntryIterator<K,V> extends TMIterator<K,Map.Entry<K,V>> {
+        private final Cursor<K, VersionedValue> cursor;
 
-        public EntryIterator(TransactionMap<K, ?> transactionMap, K from, K to) {
-            super(transactionMap, from, to, false);
+        UncommittedIterator(TransactionMap<K, ?> transactionMap, K from, K to, boolean forEntries) {
+            super(forEntries);
+            Transaction transaction = transactionMap.getTransaction();
+            this.transactionId = transaction.transactionId;
+            this.cursor = new Cursor<>(transactionMap.map.flushAndGetRoot().root, from, to);
+            fetchNext();
         }
 
         @Override
-        @SuppressWarnings("unchecked")
-        protected Map.Entry<K, V> registerCurrent(K key, VersionedValue data) {
-            return new AbstractMap.SimpleImmutableEntry<>(key, (V) data.getCurrentValue());
+        void fetchNext() {
+            while (cursor.hasNext()) {
+                K key = cursor.next();
+                VersionedValue data = cursor.getValue();
+                if (data != null) {
+                    Object currentValue = data.getCurrentValue();
+                    if (currentValue != null
+                            || transactionId != TransactionStore.getTransactionId(data.getOperationId())) {
+                       registerCurrent(key, currentValue);
+                       return;
+                    }
```

Comment:
This place is suspicious. It's not mine, the similar logic is here in the `master` branch too.

The `currentValue` can really be `null` here? If so, it should be excluded from the result instead; such entry is not valid, and if a key set is used this key is not expected here.

Reply:
Or such behavior is intentional? It is used only for unique indexes.